library/ssl_tls12_server.c

```diff
@@ -1739,16 +1649,26 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
      */
     if( sig_hash_alg_ext_present == 0 )
     {
-        uint16_t sig_algs[] = { MBEDTLS_SSL_SIG_ALG(MBEDTLS_SSL_HASH_SHA1) };
-        mbedtls_md_type_t md_default = MBEDTLS_MD_NONE;
-        for( i = 0; i < sizeof( sig_algs ) / sizeof( sig_algs[0] ) ; i++ )
+        uint16_t *set = ssl->handshake->received_sig_algs;
+        const uint16_t sig_algs[] = {
+            MBEDTLS_SSL_SIG_ALG_SET( MBEDTLS_SSL_HASH_SHA1 )
+            MBEDTLS_SSL_SIG_ALG( MBEDTLS_SSL_SIG_ANON, MBEDTLS_SSL_HASH_NONE )
+        };
+        size_t count = sizeof( sig_algs ) / sizeof( sig_algs[0] );
+
+        if( count <= MBEDTLS_RECEIVED_SIG_ALGS_SIZE )
```

Comment:
This bit looks pretty weird to my eyes, for a few of reasons:
1) This is doing some logic depending on whether a compile-time constant array is larger than a compile-time macro, so this check will always go one way (the compiler optimization pass with catch it, which is nice).
2) This makes sense if we want to be able to handle whatever someone puts in `sig_algs`, but then the special-case handling has a not-very-clear requirement that the `0x0` signature be the last element, so we can't even handle that case
3) In the special-case, we might not copy some algs from `sig_algs` to `set`, but we don't make the user aware that some aren't being copied.

If my understanding is correct, ideally we'd just treat count >= MBEDTLS_RECEIVED_SIG_ALGS_SIZE as a compile-time error (since that way we're not silently dropping algs). If we can't have that be a compile-time error, I can see it being a runtime error potentially rather than removing algs transparently.

I'd also suggest that instead of the special-case being to copy the last element of `sig_algs` to the last slot of `set`, we should just set the last slot of set to be the `0x0` (potentially `MBEDTLS_TLS_SIG_NONE` from above`). This prevent errors where someone adds another element in the wrong order.